File: ALODWebUtility/Common/ControlValues.vb
Imports System.Reflection
Imports System.Web.UI
Imports System.Web.UI.HtmlControls
Imports System.Web.UI.WebControls

Namespace Common

    Public Class ControlValues

#Region "ControlValues GetMethods"

        Public Shared Function GetAttribute(ByVal ctl As Object, ByVal TypeString As String, ByVal AttrName As String) As String
            Dim ret As String = ""
            Select Case (TypeString)
                Case "System.Web.UI.WebControls.ListBox"
                    ret = CType(ctl, ListBox).Attributes(AttrName)
                Case "System.Web.UI.WebControls.CheckBoxList"
                    ret = CType(ctl, CheckBoxList).Attributes(AttrName)
                Case "System.Web.UI.WebControls.DropDownList"
                    ret = CType(ctl, DropDownList).Attributes(AttrName)
                Case "System.Web.UI.WebControls.CheckBox"
                    ret = CType(ctl, CheckBox).Attributes(AttrName)
                Case "System.Web.UI.WebControls.TextBox"
                    ret = CType(ctl, TextBox).Attributes(AttrName)
                Case "System.Web.UI.WebControls.GridView"
                    ret = CType(ctl, GridView).Attributes(AttrName)
                Case "System.Web.UI.WebControls.RadioButtonList"
                    ret = CType(ctl, RadioButtonList).Attributes(AttrName)
                Case "System.Web.UI.WebControls.RadioButton"
                    ret = CType(ctl, RadioButton).Attributes(AttrName)
                Case "System.Web.UI.HtmlControls.HtmlInputHidden"
                    ret = CType(ctl, HtmlInputHidden).Attributes(AttrName)

            End Select
            Return ret

        End Function

        Public Shared Function GetControls(ByVal ctls As System.Web.UI.ControlCollection) As String
            Dim sb As New System.Text.StringBuilder(1024)
            Dim dc As ModControl
            Dim TypeString As String

            For Each ctl As Web.UI.Control In ctls
                If ctl.HasControls Then
                    ' Recurse into any "Grouping" controls
                    sb.Append(GetControls(ctl.Controls))
                End If

                ' Get Control Definition

                TypeString = ctl.GetType().ToString()
                dc = ControlLoad.mControlList.GetByType(TypeString)
                If Not (dc Is Nothing) Then
                    If dc.IsMultiList And dc.PropertyToCheck = String.Empty Then
                        If TypeString = "System.Web.UI.WebControls.ListBox" Then
                            sb.Append(GetMultiListValues(CType(ctl, ListBox)))
                        ElseIf TypeString = "System.Web.UI.WebControls.CheckBoxList" Then
                            sb.Append(GetMultiListValues(CType(ctl, CheckBoxList)))
                        ElseIf TypeString = "System.Web.UI.WebControls.RadioButtonList" Then
                            sb.Append(GetMultiListValues(CType(ctl, RadioButtonList)))

                        End If
                    ElseIf dc.IsMultiList And dc.PropertyToCheck <> String.Empty Then
                        If GetControlValue(ctl, dc.PropertyToCheck) = dc.PropertyToCheckValue Then
                            sb.Append(GetControlValue(ctl, dc.Property1))
                        Else
                            sb.Append(GetMultiListValues(CType(ctl, ListBox)))
                        End If
                    ElseIf dc.IsMultiList = False And dc.PropertyToCheck <> String.Empty Then
                        If GetControlValue(ctl, dc.PropertyToCheck) = dc.PropertyToCheckValue Then
                            sb.Append(GetControlValue(ctl, dc.Property1))
                        Else
                            sb.Append(GetControlValue(ctl, dc.Property2))
                        End If
                    ElseIf dc.IsMultiList = False And dc.Property2 <> String.Empty Then
                        sb.Append(GetControlValue(ctl, dc.Property1))
                        sb.Append(GetControlValue(ctl, dc.Property2))
                    Else
                        sb.Append(GetControlValue(ctl, dc.Property1))
                    End If
                End If
            Next

            Return sb.ToString()
        End Function

        Public Shared Function GetControlValue(ByVal ctl As Control, ByVal PropertyName As String) As String
            Dim t As Type
            Dim value As String = ""
            t = ctl.GetType()
            Dim TypeString As String = t.ToString()
            If (TypeString = "System.Web.UI.WebControls.DropDownList") Then
                'Customized to get the Selected Item Text instead of Selected Index
                Dim ddl As DropDownList
                ddl = CType(ctl, DropDownList)

                If ddl.SelectedIndex >= 0 Then
                    value = ddl.SelectedItem.Text
                End If
            Else
                ' Use Reflection to get the property value
                value = Convert.ToString(t.InvokeMember(PropertyName, BindingFlags.GetProperty, Nothing, ctl, Nothing))
            End If

            Return value
        End Function

        Public Shared Function GetKeyValues(ByVal ctls As System.Web.UI.ControlCollection, ByRef lstPgCrtls As lodControlList, ByVal parentAttr As String) As String

            Dim sb As New System.Text.StringBuilder(1024)
            Dim dc As ModControl
            Dim valString As String = String.Empty
            Dim sectionString As String = String.Empty
            Dim nameString As String = String.Empty
            Dim fieldString As String = String.Empty

            Dim TypeString As String = String.Empty

            For Each ctl As Control In ctls
                TypeString = ctl.GetType().ToString()
                If ctl.HasControls Then
                    parentAttr = GetAttribute(ctl, TypeString, "Section")

                    If (TypeString <> "System.Web.UI.WebControls.CheckBoxList" AndAlso TypeString <> "System.Web.UI.WebControls.RadioButtonList") Then 'This is to stop it for recursing for any child controls for CheckBox list since we take care of that in getting multilist values
                        ' Recurse into any "Grouping" controls ,pass the parentAttribute
                        sb.Append(GetKeyValues(ctl.Controls, lstPgCrtls, parentAttr))
                    End If
                End If

                ' Get Control Definition
                dc = ControlLoad.mControlList.GetByType(TypeString)

                If Not (dc Is Nothing) Then
                    If dc.IsMultiList Then
                        If (dc.PropertyToCheck = String.Empty) Then
                            If TypeString = "System.Web.UI.WebControls.ListBox" Then
                                valString = GetMultiListValues(CType(ctl, ListBox))
                                sb.Append(valString)
                            ElseIf TypeString = "System.Web.UI.WebControls.CheckBoxList" Then
                                valString = GetMultiListValues(CType(ctl, CheckBoxList))
                                sb.Append(valString)
                            ElseIf TypeString = "System.Web.UI.WebControls.RadioButtonList" Then
                                valString = GetMultiListValues(CType(ctl, RadioButtonList))
                                sb.Append(valString)
                            End If
                        Else
                            'Case for ListBox for single and multiple selection for values
                            If GetControlValue(ctl, dc.PropertyToCheck) = dc.PropertyToCheckValue Then
                                valString = GetControlValue(ctl, dc.Property1)
                                sb.Append(valString)
                            Else
                                valString = GetMultiListValues(CType(ctl, ListBox))
                                sb.Append(valString)
                            End If
                        End If
                    Else
                        valString = GetControlValue(ctl, dc.Property1)
                        sb.Append(valString)
                    End If

                    If (parentAttr = "") Then
                        sectionString = GetAttribute(ctl, TypeString, "Section")
                    Else
                        sectionString = parentAttr
                    End If

                    fieldString = GetAttribute(ctl, TypeString, "Field")

                    lstPgCrtls.Add(New lodControl(ctl.ClientID, valString, sectionString, fieldString))
                End If

            Next
            Return sb.ToString()
        End Function

        Public Shared Function GetMultiListValues(ByVal ctl As Object) As String

            Dim sb As New System.Text.StringBuilder(200)
            Dim lstItem As ListItem
            Dim i As Integer
            ' Build a concatenated list of selected indices
            For Each lstItem In ctl.Items
                If lstItem.Selected = True Then
                    i = ctl.Items.IndexOf(lstItem)
                    sb.Append(lstItem.Text + ",")
                End If
            Next

            If (sb.Length > 0) Then
                sb = sb.Remove(sb.Length - 1, 1)
            End If

            Return sb.ToString()
        End Function

#End Region

#Region "Static Functions to be called from Pages "

        'Following Static Code is used to access Values
        '**************************************************
        'Code to get Hash Values
        Public Shared Function GetInitialHash(ByVal ctls As System.Web.UI.ControlCollection) As Integer
            ' Set the Initial Hash Value
            Dim InitialHash As Integer
            InitialHash = GetControls(ctls).GetHashCode()
            Return InitialHash
        End Function

        Public Shared Function HasChanged(ByVal ctls As System.Web.UI.ControlCollection, ByVal InitialHash As String) As Boolean

            If (InitialHash = "") Then Return False 'There was no initial Value so no mods

            Dim newHash As Integer
            ' Get the Ending Hash Value
            newHash = GetControls(ctls).GetHashCode()
            Return (Int32.Parse(InitialHash) <> newHash)
        End Function

        '**************************************************

#End Region

    End Class

End Namespace

File: ALODWebUtility.CS/Common/ControlValues.cs
using System;
using System.Reflection;
using System.Web.UI;
using System.Web.UI.HtmlControls;
using System.Web.UI.WebControls;
using System.Text;

namespace ALODWebUtility.Common
{
    public class ControlValues
    {
        #region ControlValues GetMethods

        public static string GetAttribute(object ctl, string TypeString, string AttrName)
        {
            string ret = "";
            switch (TypeString)
            {
                case "System.Web.UI.WebControls.ListBox":
                    ret = ((ListBox)ctl).Attributes[AttrName];
                    break;
                case "System.Web.UI.WebControls.CheckBoxList":
                    ret = ((CheckBoxList)ctl).Attributes[AttrName];
                    break;
                case "System.Web.UI.WebControls.DropDownList":
                    ret = ((DropDownList)ctl).Attributes[AttrName];
                    break;
                case "System.Web.UI.WebControls.CheckBox":
                    ret = ((CheckBox)ctl).Attributes[AttrName];
                    break;
                case "System.Web.UI.WebControls.TextBox":
                    ret = ((TextBox)ctl).Attributes[AttrName];
                    break;
                case "System.Web.UI.WebControls.GridView":
                    ret = ((GridView)ctl).Attributes[AttrName];
                    break;
                case "System.Web.UI.WebControls.RadioButtonList":
                    ret = ((RadioButtonList)ctl).Attributes[AttrName];
                    break;
                case "System.Web.UI.WebControls.RadioButton":
                    ret = ((RadioButton)ctl).Attributes[AttrName];
                    break;
                case "System.Web.UI.HtmlControls.HtmlInputHidden":
                    ret = ((HtmlInputHidden)ctl).Attributes[AttrName];
                    break;
            }
            return ret;
        }

        public static string GetControls(System.Web.UI.ControlCollection ctls)
        {
            StringBuilder sb = new StringBuilder(1024);
            ModControl dc;
            string TypeString;

            foreach (System.Web.UI.Control ctl in ctls)
            {
                if (ctl.HasControls())
                {
                    // Recurse into any "Grouping" controls
                    sb.Append(GetControls(ctl.Controls));
                }

                // Get Control Definition

                TypeString = ctl.GetType().ToString();
                dc = ControlLoad.mControlList.GetByType(TypeString);
                if (dc != null)
                {
                    if (dc.IsMultiList && dc.PropertyToCheck == string.Empty)
                    {
                        if (TypeString == "System.Web.UI.WebControls.ListBox")
                        {
                            sb.Append(GetMultiListValues((ListBox)ctl));
                        }
                        else if (TypeString == "System.Web.UI.WebControls.CheckBoxList")
                        {
                            sb.Append(GetMultiListValues((CheckBoxList)ctl));
                        }
                        else if (TypeString == "System.Web.UI.WebControls.RadioButtonList")
                        {
                            sb.Append(GetMultiListValues((RadioButtonList)ctl));
                        }
                    }
                    else if (dc.IsMultiList && dc.PropertyToCheck != string.Empty)
                    {
                        if (GetControlValue(ctl, dc.PropertyToCheck) == dc.PropertyToCheckValue)
                        {
                            sb.Append(GetControlValue(ctl, dc.Property1));
                        }
                        else
                        {
                            sb.Append(GetMultiListValues((ListBox)ctl));
                        }
                    }
                    else if (!dc.IsMultiList && dc.PropertyToCheck != string.Empty)
                    {
                        if (GetControlValue(ctl, dc.PropertyToCheck) == dc.PropertyToCheckValue)
                        {
                            sb.Append(GetControlValue(ctl, dc.Property1));
                        }
                        else
                        {
                            sb.Append(GetControlValue(ctl, dc.Property2));
                        }
                    }
                    else if (!dc.IsMultiList && dc.Property2 != string.Empty)
                    {
                        sb.Append(GetControlValue(ctl, dc.Property1));
                        sb.Append(GetControlValue(ctl, dc.Property2));
                    }
                    else
                    {
                        sb.Append(GetControlValue(ctl, dc.Property1));
                    }
                }
            }

            return sb.ToString();
        }

        public static string GetControlValue(Control ctl, string PropertyName)
        {
            Type t;
            string value = "";
            t = ctl.GetType();
            string TypeString = t.ToString();
            if (TypeString == "System.Web.UI.WebControls.DropDownList")
            {
                //Customized to get the Selected Item Text instead of Selected Index
                DropDownList ddl;
                ddl = (DropDownList)ctl;

                if (ddl.SelectedIndex >= 0)
                {
                    value = ddl.SelectedItem.Text;
                }
            }
            else
            {
                // Use Reflection to get the property value
                value = Convert.ToString(t.InvokeMember(PropertyName, BindingFlags.GetProperty, null, ctl, null));
            }

            return value;
        }

        public static string GetKeyValues(System.Web.UI.ControlCollection ctls, ref lodControlList lstPgCrtls, string parentAttr)
        {
            StringBuilder sb = new StringBuilder(1024);
            ModControl dc;
            string valString = string.Empty;
            string sectionString = string.Empty;
            string nameString = string.Empty;
            string fieldString = string.Empty;

            string TypeString = string.Empty;

            foreach (Control ctl in ctls)
            {
                TypeString = ctl.GetType().ToString();
                if (ctl.HasControls())
                {
                    parentAttr = GetAttribute(ctl, TypeString, "Section");

                    if (TypeString != "System.Web.UI.WebControls.CheckBoxList" && TypeString != "System.Web.UI.WebControls.RadioButtonList")
                    { //This is to stop it for recursing for any child controls for CheckBox list since we take care of that in getting multilist values
                        // Recurse into any "Grouping" controls ,pass the parentAttribute
                        sb.Append(GetKeyValues(ctl.Controls, ref lstPgCrtls, parentAttr));
                    }
                }

                // Get Control Definition
                dc = ControlLoad.mControlList.GetByType(TypeString);

                if (dc != null)
                {
                    if (dc.IsMultiList)
                    {
                        if (dc.PropertyToCheck == string.Empty)
                        {
                            if (TypeString == "System.Web.UI.WebControls.ListBox")
                            {
                                valString = GetMultiListValues((ListBox)ctl);
                                sb.Append(valString);
                            }
                            else if (TypeString == "System.Web.UI.WebControls.CheckBoxList")
                            {
                                valString = GetMultiListValues((CheckBoxList)ctl);
                                sb.Append(valString);
                            }
                            else if (TypeString == "System.Web.UI.WebControls.RadioButtonList")
                            {
                                valString = GetMultiListValues((RadioButtonList)ctl);
                                sb.Append(valString);
                            }
                        }
                        else
                        {
                            //Case for ListBox for single and multiple selection for values
                            if (GetControlValue(ctl, dc.PropertyToCheck) == dc.PropertyToCheckValue)
                            {
                                valString = GetControlValue(ctl, dc.Property1);
                                sb.Append(valString);
                            }
                            else
                            {
                                valString = GetMultiListValues((ListBox)ctl);
                                sb.Append(valString);
                            }
                        }
                    }
                    else
                    {
                        valString = GetControlValue(ctl, dc.Property1);
                        sb.Append(valString);
                    }

                    if (parentAttr == "")
                    {
                        sectionString = GetAttribute(ctl, TypeString, "Section");
                    }
                    else
                    {
                        sectionString = parentAttr;
                    }

                    fieldString = GetAttribute(ctl, TypeString, "Field");

                    lstPgCrtls.Add(new lodControl(ctl.ClientID, valString, sectionString, fieldString));
                }

            }
            return sb.ToString();
        }

        public static string GetMultiListValues(object ctl)
        {
            StringBuilder sb = new StringBuilder(200);
            ListItem lstItem;
            int i;
            // Build a concatenated list of selected indices
            // ctl is Object, need to cast to access Items. 
            // ListBox, CheckBoxList, RadioButtonList all have Items property but they don't share a common interface with Items property in .NET 4.8 easily accessible without casting or dynamic.
            // But here we can use dynamic or reflection, or cast based on type.
            // The VB code relied on late binding or implicit interface.
            // I'll use dynamic to avoid repetitive casting if possible, or just cast to ListControl which is the base class for all these.
            
            ListControl listControl = ctl as ListControl;
            if (listControl != null)
            {
                foreach (ListItem item in listControl.Items)
                {
                    if (item.Selected)
                    {
                        i = listControl.Items.IndexOf(item);
                        sb.Append(item.Text + ",");
                    }
                }
            }

            if (sb.Length > 0)
            {
                sb.Remove(sb.Length - 1, 1);
            }

            return sb.ToString();
        }

        #endregion

        #region Static Functions to be called from Pages

        //Following Static Code is used to access Values
        //**************************************************
        //Code to get Hash Values
        public static int GetInitialHash(System.Web.UI.ControlCollection ctls)
        {
            // Set the Initial Hash Value
            int InitialHash;
            InitialHash = GetControls(ctls).GetHashCode();
            return InitialHash;
        }

        public static bool HasChanged(System.Web.UI.ControlCollection ctls, string InitialHash)
        {
            if (InitialHash == "") return false; //There was no initial Value so no mods

            int newHash;
            // Get the Ending Hash Value
            newHash = GetControls(ctls).GetHashCode();
            return (int.Parse(InitialHash) != newHash);
        }

        //**************************************************

        #endregion
    }
}
